datastore/protobuf/datastore.proto

```diff
@@ -140,13 +152,57 @@ message PutObsResponse {
 //---------------------------------------------------------------------------
 
 message GetObsRequest {
+  // --- BEGIN non-string metadata -------------------------
+
+  // temporal search
   TimeInterval interval = 1; // only return observations in this time range
+
+  // spatial search
   Polygon inside = 2; // if specified, only return observations in this area
-  repeated string platforms = 3; // if specified, only return observations matching any of these platform patterns
-  repeated string standard_names = 4 [json_name = "standard_names"]; // if specified, only return observations matching any of these standard names
-  repeated string instruments = 5; // if specified, only return observations matching any of these instruments
-  repeated string processing_levels = 6 [json_name = "processing_levels"]; // if specified, only return observations matching any of these processing levels
-  // TODO: add search filters for other metadata
+
+  // search wrt. TSMetadata.links
+  // TODO - needs special handling
+
+  // --- END non-string metadata -------------------------
+
+
+  // --- BEGIN string metadata (handleable with reflection) -------------------------
+  //
+  //   - field names must correspond exactly with string field names in TSMetadata or ObsMetadata
+  //
+  //   - if the field F is specified (where F is for example 'platform'), only observations
+  //     matching at least one these values for F will be returned
+  //
+  //   - if the field F is not specified, filtering on F is effectively disabled
+
+  repeated string version = 3;
+  repeated string type = 4;
+  repeated string title = 5;
+  repeated string summary = 6;
+  repeated string keywords = 7;
+  repeated string keywords_vocabulary = 8 [json_name = "keywords_vocabulary"];
+  repeated string license = 9;
+  repeated string conventions = 10;
+  repeated string naming_authority = 11 [json_name = "naming_authority"];
+  repeated string creator_type = 12 [json_name = "creator_type"];
+  repeated string creator_name = 13 [json_name = "creator_name"];
+  repeated string creator_email = 14 [json_name = "creator_email"];
+  repeated string creator_url = 15 [json_name = "creator_url"];
+  repeated string institution = 16;
+  repeated string project = 17;
+  repeated string source = 18;
+  repeated string platform = 19;
+  repeated string platform_vocabulary = 20 [json_name = "platform_vocabulary"];
+  repeated string standard_name = 21 [json_name = "standard_name"];
+  repeated string unit = 22;
+  repeated string instrument = 23;
+  repeated string instrument_vocabulary = 24 [json_name = "instrument_vocabulary"];
+  repeated string id = 25;
+  repeated string data_id = 26 [json_name = "data_id"];
+  repeated string history = 27;
+  repeated string metadata_id = 28 [json_name = "metadata_id"];
+  repeated string processing_level = 29 [json_name = "processing_level"];
```

Comment:
Can we do something like:
`map<string, repeated string> filters = 3;`

Where the key is the metadata field you want to filter on, and the value is the list of values you want to match on?

Reply:
You mean avoiding explicit references in datastore.proto too? Interesting idea. Not sure if associative arrays (maps) are supported in proto files. If yes, I guess we would need to also change the database schema in the same fashion, no? So we would end up with two levels of attributes at all levels: 1) static/explicit attributes (like obs time, geo position, etc.), and 2) dynamic/implicit attributes. On the plus side, it would be easier to alter the set of metadata at run time (no need for recompilation / redeployment - the datastore+api would simply adapt automatically to whatever the ingestor sends. On the minus side ... well I'm not sure. Maybe performance will suffer a little bit? And maybe it will be a bit trickier to support for non-trivial types (like the links attribute)?

Reply:
Sorry, I guess I was too quick and misunderstood a bit. You mean to use a dynamic/implicit scheme only for the `GetObsRequest` message! That makes good sense, and I'll look into ways to implement that (if necessary using a two-level array as a poor man's associative array etc.). Using dynamic/implicit naming for the whole shebang could also have its merits, but maybe try that at a later point?

Reply:
Yes indeed, I meant just for `GetObsRequest`.

Although the idea of dynamic/implicit attributes in the DB is interesting as well. I guess you could say it means using a narrow table approach for the metadata as well? Anyway, probably good to leave this second part for later.